File: version2/python/read_energy_force_new.py
import os,sys
import numpy as np
from trans_rot_coords import distance
from grids_structures_general import DS
import gzip


AU2KCAL = 23.0605*27.2116

class atom():
    def __init__(self, line):
        line = line.split()
        self.name = line[0]
        self.x = [float(line[2]), float(line[3]), float(line[4])]

class molecule():
    def __init__(self):
        self.natoms = 0
        self.atoms = []

    def addatom(self, line):
        self.atoms.append(atom(line))
        self.natoms += 1

    def set_energy(self, e): self.E = e


class pair_configuration():
    def __init__(self, n1, n2):
        self.n1 = n1
        self.n2 = n2
        self.fmole1 = []
        self.xmole2 = []
        self.E = 0.0
        pass

    def set_energy(self, e):
        self.E = e

    def set_force(self, ff):
        self.Fx = ff[0]
        self.Fy = ff[1]
        self.Fz = ff[2]

    def set_torque(self,tq):
        self.Tx = tq[0]
        self.Ty = tq[1]
        self.Tz = tq[2]

    def add_atom_in_mole2(self, line):
        line = line.split()
        temp = [float(line[2]),float(line[3]),float(line[4])]
        self.xmole2.append( temp )

    def get_prop(self, name, w):
        if name=='E': return self.E
        elif name == 'Fx': return self.Fx
        elif name == 'Fy': return self.Fy
        elif name == 'Fz': return self.Fz
        elif name == 'Tx': return self.Tx
        elif name == 'Ty': return self.Ty
        elif name == 'Tz': return self.Tz
        else: return 'NONE'

    def set_name(self, name):
        self.name = name

class energy_force_data():
    def __init__(self, Iname, FragType):
        self.filename = Iname
        self.FT = FragType

        self.allcfg = []
        for i in range(DS[self.FT].nDist):
            self.allcfg.append([])
            for j in range(DS[self.FT].nGrid):
                self.allcfg[i].append([])
                for m in range(DS[self.FT].nConf[i]):
                    self.allcfg[i][j].append([])
                    for k in range(DS[self.FT].nNorm[i]):
                        self.allcfg[i][j][m].append('None')

        self.natoms1 = DS[self.FT].n1
        self.natoms2 = DS[self.FT].n2

        self.ReadFile()

    def ReadFile(self):
        datafile = gzip.open( self.filename, 'r' )
        ## Read Mole 1 (config No.-1):
        while 1:
            line = datafile.readline()
            if line=='':
                raise Exception, "No data found."

            if line.find('# No.')==0:

                line = line.split()
                if line[-1]!='-1':
                    raise Exception, "Mole 1 not found."

                for ii in range(3): line = datafile.readline()
                ## energy of mole 1:
                line = line.split()
                self.mole1 = molecule()
                self.mole1.set_energy( float(line[1]) )
                for ii in range(3): line = datafile.readline()
                ## coords of mole 1:
                while 1:
                    line = datafile.readline()
                    if line.find('# End of No.')==0: break
                    self.mole1.addatom(line)
                line = datafile.readline()
                break

        ## Read Mole 2 (config No. 0):
        while 1:
            line = datafile.readline()
            if line=='':
                raise Exception, "No data found."
            if line.find('# No.')==0:
                line = line.split()
                if line[-1]!='0':
                    raise Exception, "Mole 2 not found."

                for ii in range(3): line = datafile.readline()
                ## energy of mole 2:
                line = line.split()
                self.mole2 = molecule()
                self.mole2.set_energy( float(line[1]) )
                for ii in range(3): line = datafile.readline()
                ## coords of mole 2:
                while 1:
                    line = datafile.readline()
                    if line.find('# End of No.')==0: break
                    self.mole2.addatom(line)
                line = datafile.readline()
                break

        self.distndx = []
        self.distname = []
        self.distnum = []
        ## Read configs:
        while 1:
            line = datafile.readline() # No.
            line = datafile.readline() # file name
            if line=='': break
            name = line.split()[1]
            d,g,c = self._parse_name(line)
            ic = c//DS[self.FT].nConf[d]
            c = c%DS[self.FT].nConf[d]
            datafile.readline()

            line = datafile.readline() # energy
            if line.find('Failed')>-1:
                for nn in range(3): datafile.readline()
                continue

            config = pair_configuration(self.natoms1, self.natoms2)
            config.set_name(name)

            # energy
            line = line.split()
            E = float(line[1])
            config.set_energy( E )
            datafile.readline()

            # force
            line = datafile.readline()
            tmp = line.strip().split()
            ff = [float(x) for x in tmp[1:4]]
            config.set_force( ff )

            # torque
            line = datafile.readline()
            tmp = line.strip().split()
            tq = [float(x) for x in tmp[1:4]]
            config.set_torque( tq )

            ## coords of mole2
            for nn in range(self.natoms2):
                line = datafile.readline()
                config.add_atom_in_mole2(line)

            datafile.readline() # End of This Config.
            datafile.readline()

            self.allcfg[d][g][c][ic] = config

        datafile.close()


    def get_energy(self, i,j,k):
        return self.allcfg[i][j][k]

    def get_prop(self, i,j,si,ni,name, w, ehigh=100.0):
        if i<0:
            return ehigh
        elif i>DS[self.FT].nDist-1:
            return 0.0
        cfg = self.allcfg[i][j][si][ni]
        if cfg=='None': return ehigh
        else:
            return self.allcfg[i][j][si][ni].get_prop(name, w)


    def _parse_name(self, line):
        line = line.split()
        line = line[1].rsplit('.',2)
        line = line[0].split('_')
        d = DS[self.FT].R_NDX.index(float(line[1][1:]))
        g = int(line[2][1:])
        c = int(line[3][1:])
        return d,g,c

File: version2/read_energy_force_new.cpp
#include <boost/algorithm/string.hpp>
#include <boost/lexical_cast.hpp>

#include "read_energy_force_new.h"

using namespace database;

Atom::Atom(std::string& line)
{
  std::vector<std::string> fields;
  boost::trim(line);
  boost::split(fields, line, boost::is_any_of(" \t"), boost::token_compress_on);

  assert(fields.size() == 5);
  m_name = fields[0];
  m_coord = {boost::lexical_cast<double>(fields[2]),
                         boost::lexical_cast<double>(fields[3]),
                         boost::lexical_cast<double>(fields[4])};
}

void Molecule::addAtom(std::string& line)
{
  m_atoms.push_back(Atom(line));
}


Config::Config():
  m_n1(0), m_n2(0), m_energy(0)
{}

Config::Config(int n1, int n2):
  m_n1(n1), m_n2(n2), m_energy(0)
{}

void Config::setEnergy(double energy)
{
  m_energy = energy;
}
void Config::setForce(std::array<double, 3> ff) {
    m_Fx = ff[0];
    m_Fy = ff[1];
    m_Fz = ff[2];
}

void Config::setTorque(std::array<double, 3> tq) {
    m_Tx = tq[0];
    m_Ty = tq[1];
    m_Tz = tq[2];
}

void Config::setName(std::string name)
{
  m_name = name;
}

double Config::get_prop(std::string name, double w)
{
  if (name == "E") return m_energy;
  else if (name == "Fx") return m_Fx;
  else if (name == "Fy") return m_Fy;
  else if (name == "Fz") return m_Fz;
  else if (name == "Tx") return m_Tx;
  else if (name == "Ty") return m_Ty;
  else if (name == "Tz") return m_Tz;
  else return 0;

  // TODO Here will not reach Now
  //int i = boost::lexical_cast<int>(name.substr(1, 3));
  //int j = boost::lexical_cast<int>(name.substr(4, 3));

  //if (w < 0)
  //{
  //  i = SwapH[i];
  //}
  //if (i < m_n1)
  //{
  //  return m_f
  //}
}

/**
 * EnergeForceDatabase
 *
 */
EnergeForceDatabase::EnergeForceDatabase(const char* filename, const char* fragtype):
  m_filename(filename), m_fragtype(fragtype)
{
  m_pDS = getDataStructure(fragtype);

  for (size_t i = 0; i < m_pDS->m_R_NDX.size(); i++) {
      m_all_config.push_back({});
      for (size_t j = 0; j < static_cast<size_t>(m_pDS->m_nGrid); j++) {
          m_all_config[i].push_back({});
          for (size_t m = 0; m < static_cast<size_t>(m_pDS->m_nConf[i]); m++) {
              m_all_config[i][j].push_back({});
              for (size_t k = 0; k < static_cast<size_t>(m_pDS->m_nNorm[i]); k++) {
                  m_all_config[i][j][m].push_back(NULL);
              }
          }
      }
  }

  m_natoms1 = m_pDS->m_n1;
  m_natoms2 = m_pDS->m_n2;
}

EnergeForceDatabase::~EnergeForceDatabase()
{
    for (size_t i = 0; i < m_pDS->m_R_NDX.size(); i++) {
      for (size_t j = 0; j < static_cast<size_t>(m_pDS->m_nGrid); j++) {
          for (size_t m = 0; m < static_cast<size_t>(m_pDS->m_nConf[i]); m++) {
              for (size_t k = 0; k < static_cast<size_t>(m_pDS->m_nNorm[i]); k++) {
                  delete m_all_config[i][j][m][k];
              }
          }
      }
  }

}

Config*& EnergeForceDatabase::at_all(int dim1, int dim2, int dim3, int dim4)
{
    if (dim1 < 0) {
        dim1 = m_pDS->m_R_NDX.size() - std::abs(dim1);
    }
    if (dim2 < 0) {
        dim2 = m_pDS->m_nGrid - std::abs(dim2);
    }
    if (dim3 < 0) {
        dim3 = m_pDS->m_nConf[dim1] - std::abs(dim3);
    }
    if (dim4 < 0) {
        dim4 = m_pDS->m_nNorm[dim1] - std::abs(dim4);
    }
    return m_all_config[dim1][dim2][dim3][dim4];
}


double EnergeForceDatabase::get_prop(int i, int j, int si, int ni, std::string name, double w, double eheigh=100.0)
{
  if (i < 0) return eheigh;
  else if (i > int(m_pDS->m_R_NDX.size()) - 1) return 0.0;

  Config*& cfg = at_all(i, j, si, ni);
  if (!cfg)
  {
    return eheigh;
  }
  else
  {
    return cfg->get_prop(name, w);
  }
}

// the ind only equal to -1 or zero
void EnergeForceDatabase::read_mole(int ind, igzstream& datafile)
{
  std::string line;
  std::getline(datafile, line);

  if (!boost::starts_with(line, "# No."))
  {
    return;
  }
  assert(ind == -1 || ind == 0);
  std::vector<std::string> fields;
  boost::split(fields, line, boost::is_any_of(" \t"), boost::token_compress_on);
  if (fields.back() != boost::lexical_cast<std::string>(ind))
  {
    std::cerr << "Mole "<< ind + 1 <<" not found \n";
    exit(EXIT_FAILURE);
  }

  // skip latter three lines
  std::getline(datafile, line);
  std::getline(datafile, line);
  std::getline(datafile, line);

  boost::split(fields, line, boost::is_any_of(" \t"), boost::token_compress_on);

  if (ind == -1)
  {
    m_mole1.m_energy = boost::lexical_cast<double>(fields[1]);
  }
  else
  {
    m_mole2.m_energy = boost::lexical_cast<double>(fields[1]);
  }
  // skip latter three lines
  std::getline(datafile, line);
  std::getline(datafile, line);
  std::getline(datafile, line);

  while (true)
  {
    std::getline(datafile, line);
    if (boost::starts_with(line, "# End of No."))
    {
      break;
    }
    if (ind == -1)
    {
      m_mole1.addAtom(line);
    }
    else
    {
      m_mole2.addAtom(line);
    }
  }
  // skip blank line
  std::getline(datafile, line);

}

void EnergeForceDatabase::read_file()
{
  igzstream datafile(m_filename.c_str());
  if ( ! datafile.good()) {
    std::cerr << "ERROR: Opening file '" << m_filename << "' failed.\n";
    exit(EXIT_FAILURE);
  }


  if (!datafile.eof())
  {
    // Read Mole 1 (config No.-1):
    read_mole(-1, datafile);
    // Read Mole 2 (config No.0):
    read_mole(0, datafile);
  }

  std::string line;
  std::vector<std::string> fields;
  // read configs
  while (!datafile.eof())
  {
    std::getline(datafile, line);  // No
    std::getline(datafile, line);  // filename

    if (line == "")
    {
      break;
    }
    std::string name = line.substr(2);
    // remove the last .inp.log from line
    line = boost::trim_right_copy_if(name, boost::is_any_of(".inp.log"));
    boost::split(fields, line , boost::is_any_of("_"), boost::token_compress_on);
    int d, g, c, ic;
    std::vector<double>& R_NDX = m_pDS->m_R_NDX;
    d = std::find(R_NDX.begin(), R_NDX.end(), boost::lexical_cast<double>(fields[1].substr(1))) - R_NDX.begin();
    g = boost::lexical_cast<double>(fields[2].substr(1));
    c = boost::lexical_cast<double>(fields[3].substr(1));

    ic = std::floor(c / m_pDS->m_nConf[d]);
    c = c % m_pDS->m_nConf[d];
    //skip blank
    std::getline(datafile, line);
    std::getline(datafile, line); // energy

    if (line.find("Failed") != std::string::npos)
    {
      // skip three line
      std::getline(datafile, line);
      std::getline(datafile, line);
      std::getline(datafile, line);
      continue;
    }
    Config* cfg = new Config(m_natoms1, m_natoms2);
    cfg->setName(name);
    boost::split(fields, line , boost::is_any_of(" "), boost::token_compress_on);

    cfg->setEnergy(boost::lexical_cast<double>(fields[1]));
    std::getline(datafile, line);

    // Force
    std::getline(datafile, line);
    boost::split(fields, line , boost::is_any_of(" "), boost::token_compress_on);
    cfg->setForce({boost::lexical_cast<double>(fields[1]),
            boost::lexical_cast<double>(fields[2]),
            boost::lexical_cast<double>(fields[3])});

    // Torque
    std::getline(datafile, line);
    boost::split(fields, line , boost::is_any_of(" "), boost::token_compress_on);
    cfg->setTorque({boost::lexical_cast<double>(fields[1]),
            boost::lexical_cast<double>(fields[2]),
            boost::lexical_cast<double>(fields[3])});

    // coords & forces of mole2
    for (int i=0; i<m_natoms2; i++)
    {
      std::getline(datafile, line);
      cfg->m_xmole2.push_back(Atom(line));
    }

    std::getline(datafile, line);
    std::getline(datafile, line);

    m_all_config[d][g][c][ic] = cfg;
  }
  datafile.close();
}


std::pair<double, double> database::weights_for_2_configs(const vector<double>& norm_vec, const vector<database::Atom> config1, const vector<database::Atom> config2, double cut)
{
  std::vector<double> vec1 = get_normal_unit(subtraction(config1[1].m_coord, config1[0].m_coord),
                                             subtraction(config1[2].m_coord, config1[0].m_coord));

  std::vector<double> vec2 = get_normal_unit(subtraction(config2[1].m_coord, config2[0].m_coord),
                                             subtraction(config2[2].m_coord, config2[0].m_coord));

  // treat vec1 as the x-axis, the normal of (vec1,vec2) as the z-axis
  std::vector<double>& vx = vec1;
  std::vector<double> vz = get_normal_unit(vec1, vec2);
  // np.cross
  std::vector<double> vy = {vz[1]*vx[2] - vz[2]*vx[1], vz[2]*vx[0]-vz[0]*vx[2], vz[0]*vx[1]-vz[1]*vx[0]};

  // project the "norm_vec" to the plane of vec1 and vec2, then use symmetry operation.
  double px = dot(norm_vec, vx);
  double py = dot(norm_vec, vy);

  double w1, w2;
  if (std::fabs(px) < cut)
  {
    if (py > 0) w2 = 1.0;
    else w2 = -1.0;
    w1 = 0.0;
  }
  else
  {
    double ang = atan(py / px);

    if (px < 0) ang += M_PI;
    else if (py < 0) ang += 2*M_PI;

    double period = M_PI;

    int nrot = ang / period;
    double rrot = ang - period * nrot;

    double ang2;
    if ((rrot - M_PI/2.0)>0.0)
    {
      ang2 = M_PI - rrot;
    }
    else
    {
      ang2 = rrot;
    }

    w2 = ang2 / (M_PI / 2);
    w1 = 1 - w2;
  }

  return {w1, w2};
}
